### ISSUE TITLE
Cgroups v2 freezer test randomly fails

### ISSUE THREAD

**Issue by YJDoc2** [containers/youki#132]
While running cargo test on main branch, cgroups v2 freezer test_set_freezer_state seems to fail randomly.
Curiously it seems some kind of race condition makes it fail? Because sometimes when running cargo test, it prints
```
test cgroups::v2::freezer::tests::test_set_freezer_state ... FAILED
```
and shows that test has failed, with a message
```
error: test failed, to rerun pass '--lib'
```
below result of all running tests,
sometimes it prints 
```
test cgroups::v1::freezer::tests::test_set_freezer_state ... ok
```
but still stops running unit and doc tests, with error message
```
error: test failed, to rerun pass '--lib'
```
and sometimes this test passes, and all test are run and pass.

Another thing I have noted is that order of tests is changes, as in probably cargo runs tests in parallel, which might play a role in this?

The test uses Arc, and the order of tests in cargo test output changes each time, so I feel this might be some issue related to concurrency and race conditions, but could be something else as well.

**Comment by YJDoc2** [comment]
@duduainankai if possible, can you take a look?

**Comment by duduainankai** [comment]
@YJDoc2 

Got, I will look into it ASAP.

Could you please give a link that test failed ?

**Comment by duduainankai** [comment]
I do find a potential problem in test: I used a same tmp dir for frezzer files between two different tests and tests would run in parallel, so files maybe overwritten which makes test failed.

**Comment by duduainankai** [comment]
And I found an interesting phenomenon: test failures info is not shown when do `cargo test` and if tests in `tty.rs` are commented it would show again.
Maybe it's a little annoying if we want to investigate tests failure reason in CI.

Failure test without commenting tests in `tty.rs`.
![image](https://user-images.githubusercontent.com/9639449/125227109-d6ea8480-e304-11eb-91e9-865f2ae397f0.png)

Failure test with commenting tests in `tty.rs`.
![image](https://user-images.githubusercontent.com/9639449/125227181-fbdef780-e304-11eb-8100-f26a29903d74.png)

cc @utam0k

**Comment by utam0k** [comment]
@duduainankai Has this issue already been solved by #133 ?

**Comment by duduainankai** [comment]
> @duduainankai Has this issue already been solved by #133 ?

@utam0k Yep, I think it's solved. 

But I think you take a look on https://github.com/containers/youki/issues/132#issuecomment-877943552 .

Maybe we should find a way to solve this (something like separate tests in tty.rs or a more elegant way)

**Comment by utam0k** [comment]
@duduainankai Sorry, I forgot to reply. Could you prepare a PR for this solution?

**Comment by duduainankai** [comment]
> @duduainankai Sorry, I forgot to reply. Could you prepare a PR for this solution?

Do you mean that we separate out tests in tty.rs to another step CI ?

**Comment by utam0k** [comment]
@duduainankai Sorry, I read that wrong. It would be nice to be able to test it with our own integration tests that we are working on
. I think this issue should be closed. What do you think?

**Comment by utam0k** [comment]
@duduainankai Sorry, I misread your sentence. Can I ask you to create an issue that shows the minimum way in which tty tests can fail?
I think we can create that issue and close this issue. What do you think?

**Comment by duduainankai** [comment]
> @duduainankai Sorry, I misread your sentence. Can I ask you to create an issue that shows the minimum way in which tty tests can fail?
> I think we can create that issue and close this issue. What do you think?

emm.. That's not what I mean. The tests in tty do not fail.

What I trying to say is: 

If we had a test eg. `test_set_freezer_state`  failed , with tty tests we can not read the failure message, and without tty tests(I comment the tests directly) the failure message show up again. You can see difference between the images I posted in comment.

And I think it's a little annoying if another issue like this one came up (a random test failure) and we want to read the failure message but get nothing.

I haven't read the code in tty so I have no idea what exactly happened (I guess it's 'eaten' by that 'console').

If you agree with it's an issue that we should handle, I can open another issue. 

And this issue can close in any way.

**Comment by utam0k** [comment]
I finally understand what you are trying to say. Thank you for the detailed explanation. You mean the messages under the `failures:` section are being eaten by the tty?
> If we had a test eg. test_set_freezer_state failed , with tty tests we can not read the failure message, and without tty tests(I comment the tests directly) the failure message show up again. You can see difference between the images I posted in comment.

Can I ask you to make an issue about this problem?
> If you agree with it's an issue that we should handle, I can open another issue.

**Comment by duduainankai** [comment]
> I finally understand what you are trying to say. Thank you for the detailed explanation. Do you mean the messages under the `failures:` section is being eaten by the tty?
> 
> > If we had a test eg. test_set_freezer_state failed , with tty tests we can not read the failure message, and without tty tests(I comment the tests directly) the failure message show up again. You can see difference between the images I posted in comment.
> 
> Can I ask you to make an issue about this problem?
> 
> > If you agree with it's an issue that we should handle, I can open another issue.

haha.  Sorry about my poor explanation. :mask: 

I will make another issue.
